Add custom SOP URLs to enmasse alert mails; Fix alertmanager slack vars

grdryn opened the pull request:
JIRA links:
https://issues.redhat.com/browse/INTLY-8385
https://issues.redhat.com/browse/INTLY-8386

This change adds custom sop_url info to 5 alerts. It also ensures that
2 of those alerts get routed to the critical receiver in AlertManager,
even though the severity label is set to warning.

See the first alert in the following example (the second alert is unrelated, but useful to see that the sop_url annotation is the same format):

![Screenshot from 2020-06-21 17-40-47](https://user-images.githubusercontent.com/442386/85230394-9f522700-b3e7-11ea-86ee-690d58f83053.png)

There's also a fix for an issue with how the slack properties are
configured, which meant that sometimes install/upgrade would fail when
you didn't specify those vars, but did specify other unrelated vars
for building the main alertmanager config.

## Verification Steps
As the verifier of the PR the following process should be done:

### Installation Verification
- Ensure the author of the PR has attached a log of the installation run from his branch to the jira or pr and check that it exited as expected.
- Verify the fresh installation is correct on cluster provided by PR author

Install log: https://gist.github.com/grdryn/e56350f9ad74884f1f28d14252c9bdce/raw/8a1f8a0be58eee00fb6e3ddf8b87aacf8f855de2/integreatly-install.log

### Upgrade Verification
- After installation verification, notify the PR author to begin an upgrade on their cluster
- Ensure the developer of the PR has attached a log of the upgrade run from his branch to the jira or pr and check that it exited as expected.
- If possible, look at the tasks that ran and see they match the PR
- Verify the upgrade is correct on cluster provided by PR author

Upgrade log: https://gist.github.com/grdryn/e56350f9ad74884f1f28d14252c9bdce/raw/8a1f8a0be58eee00fb6e3ddf8b87aacf8f855de2/integreatly-upgrade.log

<!--
Add the steps required to check this change. Following an example.

1. Go to `XX >> YY >> SS`
2. Create a new item `N` with the info `X`
3. Try to edit this item
4. Check if in the left menu the feature X is not so long present.
-->

## Checklist
<!-- If there is an upgrade required, either outline the steps to test it or link to the issue for the upgrade -->

- [ ] Updated [Changelog](https://github.com/integr8ly/installation/blob/master/CHANGELOG.md)
- [x] Tested Installation
- [x] Tested Uninstallation
- [x] Tested or Created follow on for Upgrade

Is an upgrade task required and are there additional steps needed to test this?
- [ ] Yes
- [ ] No
- [x] i added an upgrade task to the upgrade playbook, and I added an upgrade log from a successful run

## NOTE
* I don't actually know what needs to go into the SOPs for these alerts, but so I've just created a skeleton of each of the files referenced in this PR: https://github.com/RHCloudServices/integreatly-help/pull/564
* I don't know how to trigger all of these alerts, so I haven't been able to verify that the two warning severity alerts are correctly being routed to the critical receiver now
* These changes, once accepted, will need to be cherry-picked to the 1.7 branch

grdryn commented:
/cc @jessesarn @david-martin @alanmoran @ciaran-byrne 

Please add others you feel should review.

alanmoran commented:
@grdryn sorry for the delay in reviewing. Shouldn't these SOPs live in /alerts (https://github.com/RHCloudServices/integreatly-help/tree/master/sops/alerts)

alanmoran commented on roles/middleware_monitoring_config/defaults/main.yml:
```diff
@@ -50,6 +50,13 @@ dms_webhook_url: "{{ deadmanssnitch_url | default('') }}"
-alertmanager_slack_api_url: "{{ alertmanager_slack_api_url | default('') }}"
-alertmanager_slack_default_channel: "{{ alertmanager_slack_default_channel | default('') }}"
-
+alertmanager_slack_api_url: ''
```
@pepedocs Can you have a look at this? I'm fairly sure CS SRE inject an `alertmanager_slack_api_url ` variable during these runs?

grdryn replied:
See [this thread](https://chat.google.com/room/AAAApI2jS20/H-tfsJ7ArT0) for more info. Whether CSSRE do or don't pass it shouldn't matter here. This change fixes the fact that when I _don't_ pass it, but _do_ pass an unrelated `sendgrid_api_key` variable, then upgrade fails. If CSSRE pass a value for this `alertmanager_slack_api_url`, then it will override this empty default.

pepedocs replied:
@alanmoran Yes that variable is suppose to allow CSSRE (or any users) to inject an api for alertmanager slack integration.

@grdryn If ansible merge behavior is set to override vars this is ok, otherwise this will not work. I'm talking about this - https://docs.ansible.com/ansible/2.3/intro_configuration.html#hash-behaviour

grdryn replied:
@pepedocs my understanding of that section of the docs is that it's only referring to hashes (dictionaries), and how ansible should merge them (or not); where these `alertmanager_slack_*` variables are mere strings, rather than hashes.

I believe this is a more relevant section of the docs: https://docs.ansible.com/ansible/2.3/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable

As you can see there, variables defined at this level (role defaults) have the lowest possible precedence, so if you set values for them anywhere else, that will override the empty values here.

grdryn replied:
As additional context, you'll notice that other variables here follow the other pattern that I'm changing for these vars, and I'm not changing it for those other ones. The difference, as I see it, is that those other examples are setting a variable default of "some _other_ variable, if it exists, or an empty string" -- the important part being that they're referencing a different variable, rather than this one.

E.g.:

```yaml
pd_service_key: "{{ pagerduty_integration_key | default('') }}"
```

rather than:
```yaml
pd_service_key: "{{ pd_service_key | default('') }}"
```

pepedocs replied:
@grdryn That's right it's just a mere string. Sorry wrong docu. Generally speaking I was saying we just ensure we are expecting the right precedence, thus -e will overwrite things. So there should be no problem here if this is the case.

david-martin commented on roles/middleware_monitoring_config/templates/alertmanager.yml.j2:
```diff
@@ -35,7 +38,7 @@ receivers:
     - api_url: {{ alertmanager_slack_api_url }}
       channel: {{ alertmanager_slack_default_channel }}
-      text: "message: {{ .CommonAnnotations.message }}"
+      text: "message: {{ '{{' }} .CommonAnnotations.message {{ '}}' }}"
```
@grdryn Is this escaping correct?
Is it the first pass of templating that has the `.CommonAnnotations.message` var,
or at runtime?

grdryn replied:
It's correct now, it wasn't before. It's processed by the AlertManager go-templating at runtime rather than Ansible's jinja2 templating. See [this thread](https://chat.google.com/room/AAAApI2jS20/H-tfsJ7ArT0) for more info.

jessesarn replied:
Yes, this is correct.. it is supposed to be evaluated at runtime.  We worked on this in the integreatly-support channel last week.

grdryn commented:
> @grdryn sorry for the delay in reviewing. Shouldn't these SOPs live in /alerts (https://github.com/RHCloudServices/integreatly-help/tree/master/sops/alerts)

Yes, fixed now (and the files updated in the corresponding PR to integreatly-help

jessesarn commented:
/lgtm

grdryn commented:
/hold

I've got one more thing to add here

grdryn commented:
/unhold

grdryn commented:
/retest

grdryn commented:
/cherrypick v1.7

grdryn commented:
@jessesarn @david-martin this is ready for review again now. There are two changes since last review:

* Ensure routing to critical receiver for all 5 alerts (the one that I'm able to actually trigger, ComponentHealth, didn't seem to be getting routed there by default for some reason
* Ensure that SOP URL is available in pagerduty. It's in a slightly different place than before:

![12475915_image-2020-06-23-19-08-40-823](https://user-images.githubusercontent.com/442386/85697082-63fa7580-b6d1-11ea-81bc-511353105928.png)

jessesarn commented:
Looks good from my end, @david-martin said he was going to review as well.   Deferring to him for the official LGTM.

david-martin commented:
/lgtm
Thanks @grdryn
